### DISCUSSION TITLE
Without Loading no QuerySelectorAllAsync

### DISCUSSION THREAD

**Discussion by Mephisto090** [cefsharp/CefSharp.Dom#12] [Q&A] [answered]
hey developer, :)

I'm loading the URL of a webpage through Form Show.

Now when I press a button without "Chroumium.LoadUrl" it doesn't over

 Dim DevToolsContext = Await cwbBrowser.GetDevToolsContextAsync()
 Dim Element = Await DevToolsContext.QuerySelectorAllAsync("input")
to access the website.

If I send a MsgBox after (Dim Element = Await DevToolsContext.QuerySelectorAllAsync("input")) this is no longer executed, i.e. it aborts the function without an error

Do you have any idea?

**Comment by amaitland** [comment, marked as answer]
> Now when I press a button without "Chroumium.LoadUrl" it doesn't over

What do you mean by this? What's your working code look like. 

Please provide a detailed example. Insufficient information for me to speculate.

**Reply by Mephisto090** [reply]
```
    Public Sub New()
        InitializeComponent()

        Dim UserAgent As New cUserAgent
        chromiumWebBrowser = Me.cwbBrowser

        Dim settings As New CefSettings()
        CefSharp.Cef.Initialize(settings)
    End Sub
```

The inputs are not filled with this code:
```
    Private Async Sub frmMainform_Shown(sender As Object, e As EventArgs) Handles MyBase.Shown
        chromiumWebBrowser.LoadUrl("www.example.de")
    End Sub

    Private Sub btnButton_Click(sender As Object, e As EventArgs) Handles btnButton.Click
            Await chromiumWebBrowser.WaitForInitialLoadAsync()
            Dim DevToolsContext = Await chromiumWebBrowser.GetDevToolsContextAsync()
            Dim Element = Await DevToolsContext.QuerySelectorAllAsync("input")

            For Each Elements In Element
                If Await Elements.GetAttributeValueAsync(Of String)("id") = "id1" Then
                    Await Elements.SetAttributeValueAsync("value", "Test1")
                End If
                If Await Elements.GetAttributeValueAsync(Of String)("id") = "id2" Then
                    Await Elements.SetAttributeValueAsync("value", "Test2")
                End If
            Next
    End Sub
```


The inputs are filled with this code:
```
    Private Sub btnButton_Click(sender As Object, e As EventArgs) Handles btnButton.Click
            chromiumWebBrowser.LoadUrl("www.example.de")
            Await chromiumWebBrowser.WaitForInitialLoadAsync()
            Dim DevToolsContext = Await chromiumWebBrowser.GetDevToolsContextAsync()
            Dim Element = Await DevToolsContext.QuerySelectorAllAsync("input")

            For Each Elements In Element
                If Await Elements.GetAttributeValueAsync(Of String)("id") = "id1" Then
                    Await Elements.SetAttributeValueAsync("value", "Test1")
                End If
                If Await Elements.GetAttributeValueAsync(Of String)("id") = "id2" Then
                    Await Elements.SetAttributeValueAsync("value", "Test2")
                End If
            Next
    End Sub
```

**Reply by amaitland** [reply]
Is `frmMainform_Shown` called before `btnButton_Click`. Is there are reason you are loading a url on the `Shown` event? Seems like a strange choice. Would `Load` event be a better choice?

`WaitForInitialLoadAsync` probably isn't being used correctly in this context. It will only wait for the first (initial load). If you've already loaded a web page then it will resolve immediately. 

Example still has gaps so a bit hard to tell exactly what's going on. Generally I'd go with something like calling `LoadUrlAsync` and `awaiting` that call and only enabling the button once the call has complete, thus ensuring the page has loaded before attempting to execute any javascript.

I'd be guessing you've got a problem with your timing.

**Reply by Mephisto090** [reply]
I built the loading of the website into the load event

I build the `WaitForInitialLoadAsync` right after loading a web page, so I don't need this before every access to the web page?

I tried building the `WaitForInitialLoadAsync` after loading the webpage and using `LoadUrlAsync` to load the webpage. Sadly it still does not work :(

You can also add me to Discord and I'll show you the entire source code. After that we can write a solution here in the forum.

**Reply by amaitland** [reply]
I don't use discord.

You can create a minimal example and upload it to GitHub. Keep the code to a bare minimum, just the minimum required to reproduce the problem. 

https://stackoverflow.com/help/minimal-reproducible-example has guidance

**Reply by amaitland** [reply]
Commit https://github.com/cefsharp/Puppeteer/commit/dde329e97d5e092e7ad906205443d7ab89150d91 adds a very basic `WinForms` example (ported from `CefSharp.MinimalExample`). 

Commit https://github.com/cefsharp/Puppeteer/commit/02a806426834681bb319188b9e68b96530f44e4e adds a menu item to highlight all links.

You can now work the project and use as a base to reproduce your problem. Keep your changes minimal and push them to `GitHub`.

**Reply by Mephisto090** [reply]
I managed to get an error thrown. does he help you? (Unfortunately, the error will be in German)

``
CefSharp.Puppeteer.TargetClosedException: "Protocol error(Runtime.enable): Target closed. (Page failed to process Runtime.consoleAPICalled. Der Delegat einer Instanzenmethode kann kein "this" von NULL haben..    bei System.MulticastDelegate.ThrowNullThisInDelegateToInstance()
   bei CefSharp.Puppeteer.DevToolsContext.OnConsoleAPIAsync(PageConsoleResponse message) in C:\projects\puppeteer\lib\PuppeteerSharp\DevToolsContext.cs:Zeile 2296.
   bei CefSharp.Puppeteer.DevToolsContext.<Client_MessageReceived>d__217.MoveNext() in C:\projects\puppeteer\lib\PuppeteerSharp\DevToolsContext.cs:Zeile 2097.)"
``

**Reply by amaitland** [reply]
You'll need to provide an example that reproduces the problem.

**Reply by Mephisto090** [reply]
It throws this error if the variable "`devtoolsContext`" is Nothing (`Dim devtoolsContext = Await WebBrowser.GetDevToolsContextAsync()`)

``
CefSharp.Puppeteer.TargetClosedException
  HResult=0x80131500
  Nachricht = Protocol error(Runtime.enable): Target closed. (Page failed to process Runtime.consoleAPICalled. Der Delegat einer Instanzenmethode kann kein "this" von NULL haben..    bei System.MulticastDelegate.ThrowNullThisInDelegateToInstance()
   bei CefSharp.Puppeteer.DevToolsContext.OnConsoleAPIAsync(PageConsoleResponse message) in C:\projects\puppeteer\lib\PuppeteerSharp\DevToolsContext.cs:Zeile 2296.
   bei CefSharp.Puppeteer.DevToolsContext.<Client_MessageReceived>d__217.MoveNext() in C:\projects\puppeteer\lib\PuppeteerSharp\DevToolsContext.cs:Zeile 2097.)
  Quelle = mscorlib
``

I have now created a MinimalExample. How can I set this up on GitHub?

**Reply by amaitland** [reply]
Please fork this repository, use the example as a base, keep your changes to a minimum and push them to GitHub. 

There are plenty of guides online if you are new to GitHub.

**Reply by Mephisto090** [reply]
https://github.com/Mephisto090/MinimalChrome

I added you as a collaborator.

I found 2 errors. These can be found in the CLogin class under Error 1 and Error 2

**Reply by amaitland** [reply]
> I found 2 errors. These can be found in the CLogin class under Error 1 and Error 2

One error at a time, please keep the scope to just the error you have reported here.

I need the bare minimum to ideally create a self contained unit test from. Is the example the bare minimum required to reproduce the problem? If not remove anything that's not required. 

- Remove any unnessicary classes
- remove any unnessicary code

You'll also need to provide a list of steps to reproduce the problem (basically instructions).

It's very important it's just the minimum required, otherwise I'll have to spend time removing code. I don't have time for that.

Thanks.

**Reply by Mephisto090** [reply]
My setup is really just the minimum now.

To get to the instructions. Press the "Push Me" button and see for yourself.
He will try to press the button at the top right of the website. There he will fall. 1% of the cases has already worked. If it works just restart and press again.

**Reply by amaitland** [reply]
> My setup is really just the minimum now.

Would be better if there weren't any classes, think of how you'd write a unit test. Just having the raw `API` calls in your `Form` class rather than having classes like `CWebFunctionsLibrary`. They add to the complexity and really aren't necessary to reproduce the problem. 

Call stack in `English` (I only speak `English`).

```
TargetClosedException: Protocol error(Runtime.enable): Target closed. (Page failed to process Runtime.consoleAPICalled. Delegate to an instance method cannot have null 'this'..    at System.MulticastDelegate.ThrowNullThisInDelegateToInstance()
   at CefSharp.Puppeteer.DevToolsContext.OnConsoleAPIAsync(PageConsoleResponse message) in C:\projects\puppeteer\lib\PuppeteerSharp\DevToolsContext.cs:line 2296
   at CefSharp.Puppeteer.DevToolsContext.<Client_MessageReceived>d__217.MoveNext() in C:\projects\puppeteer\lib\PuppeteerSharp\DevToolsContext.cs:line 2097)
```

The error is a little misleading. Looking at the `puppeteer` source and I think the key is at https://github.com/puppeteer/puppeteer/pull/3866/files#diff-f52aea9fb804e964b7ff5d3447d710323260033d32cbd4803dfe25c8570acc2dR511. The `DevToolsContext` is getting messages that when it doesn't have a matching `ExecutionContext`. The original port of code from `Puppeteer Sharp` doesn't have any checks for valid context, just logs an error. Commit https://github.com/cefsharp/Puppeteer/commit/9e40c363d44e00ff275d8269fea83a15da96bab8 adds a null check. You can test it out in version `1.0.38`.

I reccommend only creating one `DevToolsContext` instance per `ChromiumWebBrowser`. Reuse the same context rather than creating a new one. Perhaps `GetDevToolsContextAsync` should be renamed to `CreateDevToolsContextAsync` to make this clearer. The documentation can also be updated.

**Reply by Mephisto090** [reply]
Ok, so far it has worked great. Thank you very much.

Unfortunately, there is now the problem that if I have several such functions in one function, only the first one is executed.

Check out this class `CLogin` and the function `SetUsernameAndPassword`

First the button is pressed and then the fields are filled out. Unfortunately, he just presses the button. If I comment out the button, it writes in the first field.

**Reply by amaitland** [reply]
Reuse the same DevTools context rather than creating a new one.

If you are still having a problem then please remove all the unnessicary classes and move all the calls into the button click handler,. ,such easier to review then.

**Reply by amaitland** [reply]
If you structure your button click like a unit test then it's easy to review and can be converted into a unit test if there is an actual bug. There are plenty of tests in the source you can use as reference. One example https://github.com/cefsharp/Puppeteer/blob/main/lib/PuppeteerSharp.Tests/QuerySelectorTests/ElementHandleQuerySelectorAllTests.cs#L22

**Reply by Mephisto090** [reply]
For the clarity of the class, I would click the button and write separately in the input field. I need both functions very often and would have to write a new function for every eventuality. Theoretically you could check it as it is currently.

I will try to shorten the minimal program even further

`Reuse the same DevTools context rather than creating a new one.`

If I use the `devtoolsContext ` again, doesn't it have to download the page source code again and not return it "empty"?


it works if I read the DevToolsContext beforehand and pass it to the function.

It works:
```
DevtoolsContext = Await ChromWebBrowser.GetDevToolsContextAsync()
Await WebFunction.SetWebbrowserClick(DevtoolsContext, "strong", "class", "header-login-label", True)
```


That does not work:
`Await WebFunction.SetWebbrowserClick("strong", "class", "header-login-label", True)`

and read inside the function(`DevtoolsContext = Await ChromWebBrowser.GetDevToolsContextAsync()`)

Is there a possibility that it's also possible if I can read it out inside the function?

The problem is that if I press a button on the page, the website could change (after logging in, for example)

Can I ask what your time zone is?

**Reply by amaitland** [reply]
> If I use the devtoolsContext again, doesn't it have to download the page source code again and not return it "empty"?

What do you mean? The DevTools context is meant to be long lived, it internally tracks navigations. 

> it works if I read the DevToolsContext beforehand and pass it to the function

Sounds like everything is working as expected. You can write extension methods to encapsulate your own logic.

> The problem is that if I press a button on the page, the website could change (after logging in, for example)

This is expected. What is the problem? Internally the navigation will be tracked.

> Can I ask what your time zone is?

For what reason?

**Reply by Mephisto090** [reply]
> For what reason?

Then I can roughly know when to expect an answer and don't have to check regularly. The time difference to Germany is big. (about 10 hours) :D

> This is expected. What is the problem? Internally the navigation will be tracked.

When I reload the `DevtoolsContext` it doesn't work like before again. I don't get an error, but I also don't get anything from the "return" page. Apparently changing the page is not passed to the `DevtoolsContext`. Sometimes only a JavaScript function is triggered that outputs a text like "Wrong username..." is that also processed internally?

I will adapt the minimal example tomorrow

**Reply by amaitland** [reply]
> Then I can roughly know when to expect an answer 

I will attempt to respond to GitHub  **monthly** sponsors within 24hrs. I will respond to other queries when time permits, paid work takes priority. 

> When I reload the DevtoolsContext it doesn't work like before again

What is reloading the DevToolsContext exactly? 

> JavaScript function is triggered that outputs a text like "Wrong username..." is that also processed internally?

It's not clear what you mean by this.

> I will adapt the minimal example tomorrow

It's unlikely I will have time to look at your example this week. Maybe early next week.

**Reply by Mephisto090** [reply]
> I will attempt to respond to GitHub monthly sponsors within 24hrs. I will respond to other queries when time permits, paid work takes priority.

I think I'll buy you 2 coffees again for your great support.
Sorry for my bad English (i need the google translate) :(

> What is reloading the `DevToolsContext` exactly?

- I use the DevToolsContext and describe the input fields and press a button.
- After that, the page reloads
- After that I use the `DevToolsContext` again (with this: `DevtoolsContext = Await ChromWebBrowser.GetDevToolsContextAsync()`) and get nothing back

I have the suspicion because you wrote that the script internally notices that the page is reloading and that is not passed to `DevToolsContext`

> It's not clear what you mean by this.

When I log into a site and enter a wrong password, the site often doesn't reload but just writes "Wrong password or username"

Does the script also get this internally? As I said, the page is not reloaded in the classic sense

**Reply by amaitland** [reply]
> * I use the DevToolsContext and describe the input fields and press a button.

As per my recommendation above you should create a **single** `DevToolsContext` per `ChromiumWebBrowser` instance, not one per button click. If you need to for some reason create them in sequence then you will need to **Dispose** of the previous instance.

> * After that, the page reloads

It's hard to test as you haven't provided an actual working example. There are currently [501 passing unit tests](https://ci.appveyor.com/project/cefsharp/puppeteer/builds/42884172/tests) that were ported from `puppeteer-sharp`. So I'm fairly confident that most of the navigation scenarios are handled.

> When I log into a site and enter a wrong password, the site often doesn't reload but just writes "Wrong password or username"

Sounds like an expected use case. Shouldn't be anything special required.

> Does the script also get this internally?

Get what exactly? The source is there should you wish to review anything for yourself. 

I've **simplified** your example you can see the code at https://github.com/amaitland/Mephisto090.MinimalChrome/blob/master/MinimalChrome/Form1.vb#L41

The changes are in a single commit https://github.com/amaitland/Mephisto090.MinimalChrome/commit/8836715ee0a1bab7445efea8998ec1ad19235426 so you can review exactly was changed (you pull the changes into your branch for testing).

- I've **removed** all the unnecessary classes (which I've been repeatedly asking you to do) so the code example is clean and easy to review.
- `DevToolsContext` is now created just one on form load and disposed when the form is disposed (code for dispose is in the designer)
- Changed to use querySelector as per my previous recommendation.

I very much dislike `VB.Net` and I don't speak `German` so debugging/refactoring took way longer than it would have normally (reason I asked many many times for you to simplify your code).

**Reply by Mephisto090** [reply]
Can you tell me where the difference between `#` and `.` is? 

`Await _devtoolsContext.QuerySelectorAsync("#login")`
`Await _devtoolsContext.QuerySelectorAsync("div .header-login-button")`

**Reply by amaitland** [reply]
This is standard query selector syntax see https://developer.mozilla.org/en-US/docs/Web/API/Document_object_model/Locating_DOM_elements_using_selectors

Some basic examples see https://developer.mozilla.org/en-US/docs/Web/API/Document_object_model/Locating_DOM_elements_using_selectors#examples (This should answer your question).
